gemdaq-testing/gemsupervisor/src/common/tbutils/ThresholdScan.cc

```diff
@@ -75,34 +94,271 @@ gem::supervisor::tbutils::ThresholdScan::~ThresholdScan()
   wl_->cancel();
   wl_ = 0;
   
-  if (histo)
-    delete histo;
+  if (histo) delete histo;
   histo = 0;
 
   for (int hi = 0; hi < 128; ++hi) {
-    if (histos[hi])
-      delete histos[hi];
+    if (histos[hi]) delete histos[hi];
     histos[hi] = 0;
   }
 
-    if (outputCanvas)
-      delete outputCanvas;
-    outputCanvas = 0;
+  if (outputCanvas) delete outputCanvas;
+  outputCanvas = 0;
 
 }
 
+// State transitions
+void gem::supervisor::tbutils::ThresholdScan::configureAction(toolbox::Event::Reference e)
+  throw (toolbox::fsm::exception::Exception) {
 
-void gem::supervisor::tbutils::ThresholdScan::actionPerformed(xdata::Event& event)
-{
-  // This is called after all default configuration values have been
-  // loaded (from the XDAQ configuration file).
-  if (event.type() == "urn:xdaq-event:setDefaultValues") {
+  is_working_ = true;
+
+  latency_   = scanParams_.bag.latency;
+  nTriggers_ = confParams_.bag.nTriggers;
+  stepSize_  = scanParams_.bag.stepSize;
+  minThresh_ = scanParams_.bag.minThresh;
+  maxThresh_ = scanParams_.bag.maxThresh;
+  
+  hw_semaphore_.take();
+  vfatDevice_->setDeviceBaseNode("OptoHybrid.GEB.VFATS."+confParams_.bag.deviceName.toString());
+
+  //make sure device is not running
+  vfatDevice_->setRunMode(0);
+
+  /****unimplemented at the moment
+  if ((confParams_.bag.settingsFile.toString()).rfind("xml") != std::string::npos) {
+    LOG4CPLUS_INFO(getApplicationLogger(),"loading settings from XML file");
+    gem::supervisor::tbutils::VFAT2XMLParser::VFAT2XMLParser theParser(confParams_.bag.settingsFile.toString(),
+								       vfatDevice_);
+    theParser.parseXMLFile();
+  }
+  */
+  
+  //else {
+    LOG4CPLUS_INFO(getApplicationLogger(),"loading default settings");
+    //default settings for the frontend
+    vfatDevice_->setTriggerMode(    0x3); //set to S1 to S8
+    vfatDevice_->setCalibrationMode(0x0); //set to normal
+    vfatDevice_->setMSPolarity(     0x1); //negative
+    vfatDevice_->setCalPolarity(    0x1); //negative
+    
+    vfatDevice_->setProbeMode(        0x0);
+    vfatDevice_->setLVDSMode(         0x0);
+    vfatDevice_->setDACMode(          0x0);
+    vfatDevice_->setHitCountCycleTime(0x0); //maximum number of bits
+    
+    vfatDevice_->setHitCountMode( 0x0);
+    vfatDevice_->setMSPulseLength(0x3);
+    vfatDevice_->setInputPadMode( 0x0);
+    vfatDevice_->setTrimDACRange( 0x0);
+    vfatDevice_->setBandgapPad(   0x0);
+    vfatDevice_->sendTestPattern( 0x0);
+    
+    
+    vfatDevice_->setIPreampIn(  168);
+    vfatDevice_->setIPreampFeed(150);
+    vfatDevice_->setIPreampOut(  80);
+    vfatDevice_->setIShaper(    150);
+    vfatDevice_->setIShaperFeed(100);
+    vfatDevice_->setIComp(      120);
+
+    vfatDevice_->setLatency(latency_);
+    //}
+  
+  vfatDevice_->setVThreshold1(0-minThresh_);
```

Comment:
This methodology is wrong, unless there is some special treatment of the stepping mechanism depending on Vmin and Vmax.  We do not want the scan to only scan "symmetrically", rather, we want the scan to be able to be either negative or positive or cross the 0 boundary, i.e., we would like to specify Vmin and Vmax only, and have these correspond to the minimum and maximum DVt values.
Consider the following 6 possible scanning scenarios.  These scenarios always fix V2 (max{0,Vmax}) and step through V1 from V1 initial to V1 final (min{Vmax,0}) as we have done previously and as makes the most sense for GEMs; there are limitations to this which do not matter for our use case).
I write each line as (Vmin,Vmax): {SB initial (V1,V2)}, {desired initial (V1,V2)}
1) (-20,-5): {SB initial (20,10)} if V1 starts at 20 and steps towards 0, DVt will start at -10 and end at 10, {desired initial (20,0)}
2) (-20,0): {SB initial (20,10)} behaves just as the previous case, {(20,0)}
3) (-20,10): {SB initial (20,10)} behaves just as the previous case, {(30,10)}
4) (-10,20): {SB initial (10,5)} with V2 fixed at 5, it is not possible to get DVt = 20, {(30,20)}
5) (5,20): {SB initial (-5,-2.5)} NOT POSSIBLE to set V1 or V2 to a negative value, {(15,20)}
6) (0,20): {SB initial (0,0)} this scenario will never do anything, {(0,20)}

In general, for our threshold scan, if the Vmax is > 0, then V2 should be set to Vmax
V1 should then be computed to allow DVt to scan between Vmin and Vmax
The register we will vary will be V1.
Make sure that the scan functions as expected.  Let me know if you have more questions.